actix-settings/src/lib.rs

```diff
@@ -0,0 +1,705 @@
+
+impl<A> BasicSettings<A>
+where
+    A: de::DeserializeOwned,
+{
+    /// NOTE **DO NOT** mess with the ordering of the tables in this template.
+    ///      Especially the `[application]` table needs to be last in order
+    ///      for some tests to keep working.
+    pub(crate) const DEFAULT_TOML_TEMPLATE: &'static str = include_str!("./defaults.toml");
+
+    /// Parse an instance of `Self` from a `TOML` file located at `filepath`.
+    /// If the file doesn't exist, it is generated from the default `TOML`
+    /// template, after which the newly generated file is read in and parsed.
+    pub fn parse_toml<P>(filepath: P) -> AtResult<Self>
+    where
+        P: AsRef<Path>,
+    {
+        let filepath = filepath.as_ref();
+
+        if !filepath.exists() {
+            Self::write_toml_file(filepath)?;
+        }
+
+        let mut f = File::open(filepath)?;
+        let mut contents = String::with_capacity(f.metadata()?.len() as usize);
+        f.read_to_string(&mut contents)?;
+
+        Ok(toml::from_str::<Self>(&contents)?)
+    }
+
+    /// Parse an instance of `Self` straight from the default `TOML` template.
+    pub fn from_default_template() -> AtResult<Self> {
+        Self::from_template(Self::DEFAULT_TOML_TEMPLATE)
+    }
+
+    /// Parse an instance of `Self` straight from the default `TOML` template.
+    pub fn from_template(template: &str) -> AtResult<Self> {
+        Ok(toml::from_str(template)?)
+    }
+
+    /// Write the default `TOML` template to a new file, to be located
+    /// at `filepath`.  Return a `Error::FileExists(_)` error if a
+    /// file already exists at that location.
+    pub fn write_toml_file<P>(filepath: P) -> AtResult<()>
+    where
+        P: AsRef<Path>,
+    {
+        let filepath = filepath.as_ref();
+        let contents = Self::DEFAULT_TOML_TEMPLATE.trim();
+
+        if filepath.exists() {
+            return Err(AtError::FileExists(filepath.to_path_buf()));
+        }
+
+        let mut file = File::create(filepath)?;
+        file.write_all(contents.as_bytes())?;
+        file.flush()?;
+
+        Ok(())
+    }
+
+    pub fn override_field<F, V>(field: &mut F, value: V) -> AtResult<()>
+    where
+        F: Parse,
+        V: AsRef<str>,
+    {
+        *field = F::parse(value.as_ref())?;
+        Ok(())
+    }
+
+    pub fn override_field_with_env_var<F, N>(field: &mut F, var_name: N) -> AtResult<()>
+    where
+        F: Parse,
+        N: AsRef<str>,
+    {
+        match env::var(var_name.as_ref()) {
+            Err(env::VarError::NotPresent) => Ok((/*NOP*/)),
+            Err(var_error) => Err(AtError::from(var_error)),
+            Ok(value) => Self::override_field(field, value),
+        }
+    }
+}
+
+pub trait ApplySettings {
+    /// Apply a [`BasicSettings`] value to `self`.
+    ///
+    /// [`BasicSettings`]: ./struct.BasicSettings.html
+    #[must_use]
+    fn apply_settings<A>(self, settings: &BasicSettings<A>) -> Self
```

Comment:
Perhaps the return type should be altered to return some kind of `Result<Self, ...>`.
That way, the implementing method in the `impl ApplySettings for HttpServer` below can properly propagate its error rather than simply `unwrap()` ing it. That error can only happen if binding 1 or more addresses fails, which for valid addresses generally isn't an issue.
But technically it's still a landmine I guess.

Reply:
Yeah. Probably better to leave it for now and pencil it for a subsequent release.

Reply:
Works for me. By that time, the error should probably be a new `AtError` variant.